### PR TITLE
add a source_graph library to find the transitive deps of an entry-point

### PR DESCRIPTION
**Pull request by devoncarew**
This uses the analyzer library to parse the directives of dart files, traversing from an entry-point to all referenced sources. This will be used to send dart sources to the VM's dev FS. This is non-incremental - incrementality will be added as follow-on work.

Parsing the graph for flutter_gallery:

```
452 sources parsed in 684ms.
```

Making it incremental will drop later parses to ~ms range, and there's some additional possible optimization w/ only tokenizing the parts of the source file that we need (the directives).

@bwilkerson for review

/cc @johnmccutchan @turnidge

### CODE REVIEW

File: packages/flutter_tools/lib/src/dart/source_graph.dart [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+    // Parse any dirty sources.
+    while (sources.any((GraphSource source) => source.dirty)) {
+      for (GraphSource source in sources.where((GraphSource source) => source.dirty).toList()) {
+        source.parse();
+      }
+    }
+  }
+
+  GraphSource _getCreateSource(String resolvedPath) {
```
Comment (bwilkerson) [inline packages/flutter_tools/lib/src/dart/source_graph.dart]: Consider using `putIfAbsent`.

File: packages/flutter_tools/lib/src/dart/source_graph.dart [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    if (_sourceMap.containsKey(resolvedPath)) {
+      return _sourceMap[resolvedPath];
+    } else {
+      GraphSource source = new GraphSource(this, resolvedPath);
+      _sourceMap[resolvedPath] = source;
+      return source;
+    }
+  }
+
+  String _resolveDirectiveUri(String uriString, String basePath) {
```
Comment (bwilkerson) [inline packages/flutter_tools/lib/src/dart/source_graph.dart]: A minor observation: this duplicates some of the logic in SourceFactory. Possibly more importantly, it doesn't match the VM's rules for URI equality. I don't know what the VM's hot-patching API looks like, but if you're sending a mapping from URI to file content, there might be a subtle bug here related to relative URI's. We should talk off-line.

File: packages/flutter_tools/lib/src/dart/source_graph.dart [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+  }
+
+  void parse() {
+    dirty = false;
+    references.clear();
+
+    CompilationUnit unit = parseDirectives(getSource(), name: fullpath, suppressErrors: true);
+
+    for (Directive directive in unit.directives) {
+      if (directive is ExportDirective || directive is ImportDirective || directive is PartDirective) {
```
Comment (bwilkerson) [inline packages/flutter_tools/lib/src/dart/source_graph.dart]: If you write `if (directive is UriBasedDirective) {`, then you don't need a local variable or the implicit cast.

**Comment by bwilkerson** [comment]
lgtm

File: packages/flutter_tools/lib/src/dart/source_graph.dart (hunk 3 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/dart/source_graph.dart]: awesome - changed

File: packages/flutter_tools/lib/src/dart/source_graph.dart (hunk 1 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/dart/source_graph.dart]: updated

**Comment by johnmccutchan** [comment]
Why is it necessary to parse the source at all? I believe the following is sufficient:

Assuming your project is at $PROJECT_ROOT
- Recursively copy everything under $PROJECT_ROOT to the devfs.
- For each package referenced in the .packages:
   Recursively copy everything under $PACKAGE_LIB_DIRECTORY to the devfs

File: packages/flutter_tools/lib/src/dart/source_graph.dart (hunk 2 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/dart/source_graph.dart]: Thanks - taking a look at this now. I think we'll live with some of the duplication for a while, but I'll need to adopt some of the uri resolution logic - in particular it looks like `resolveRelativeUri()`.

**Comment by devoncarew** [comment]
> Why is it necessary to parse the source at all? I believe the following is sufficient:

Good idea. I can take a look at sending over the full directory in parallel to this PR.

**Comment by devoncarew** [comment]
Will re-cycle this one in a bit -

**Pull request closed without merging**